server/src/main/java/com/cloud/server/StatsCollector.java

```diff
@@ -1485,6 +1487,32 @@ private void storeVirtualMachineStatsInMemory(VmStatsEntry statsForCurrentIterat
         }
     }
 
+    /**
+     * Removes stats for a given virtual machine.
+     * @param vmId ID of the virtual machine to remove stats.
+     */
+    public void removeVirtualMachineStats(Long vmId) {
+        s_logger.debug(String.format("Removing stats from VM with ID: %s .",vmId));
+        _VmStats.remove(vmId);
+    }
+
+    /**
+     * Removes stats of virtual machines that are not running from memory.
+     */
+    protected void cleanUpVirtualMachineStats() {
+        List<Long> allRunningVmIds = new ArrayList<Long>();
+        for (UserVmVO vm : _userVmDao.listAllRunning()) {
```

Comment:
VM stats are captured for running VMs only
https://github.com/apache/cloudstack/blob/a78af608144f80649c409ebd933265d2d755f289/server/src/main/java/com/cloud/server/StatsCollector.java#L579

 _removeVirtualMachineStats()_  removes the vm stat if the VM is stopped/destroyed after adding to the stats, so no need for this cleanup method.

Reply:
Thanks for the review @sureshanaparti!
The `cleanUpVirtualMachineStats()` method is important to maintain consistency of the metric stats shown by ACS even when there are cases of VMs changing state unexpectedly (e.g., when a VM crashes; or when the VM is stopped directly by the hypervisor interface, etc). In these cases the `removeVirtualMachineStats()` method is not executed, so the `cleanUpVirtualMachineStats()` method does its job.